File: Algorithms & Data Structures/Data structures implementation JS/All combined.js
let array = [3, 1, 6, 8, 7, 5, 2, 4];
let sortedArray = [1, 2, 3, 4, 5, 6, 7, 8]
/*

Linear Search (normal + recursion)
Binary Search (normal + recursion)
Linked List (Single, Double, Circular)
Queue
Stack
BST

*/
class Algorithm {
    constructor() {}
    linearSearch(array, item) {
        for (let i = 0; i < array.length; i++) {
            if (array[i] == item) {
                return "Found " + item + " at position " + (i + 1);
            }
        }
        return item + " not found!";
    }
    linearRec(array, item, left, right) {
        if (left <= right) {
            if (array[left] == item) {
                return "Found " + item + " at position " + (left + 1);
            }
            if (array[right] == item) {
                return "Found " + item + " at position " + (right + 1);
            }
            return this.linearRec(array, item, left + 1, right - 1);
        }
        return item + " not found!";
    }
    binarySearch(array, item) {
        let left = 0,
            right = array.length - 1,
            found = false;
        while (left <= right && !found) {
            let mid = Math.floor((right + left) / 2);
            console.log(mid);
            if (item > array[mid]) {
                left = mid + 1;
            }
            else if (item == array[mid]) {
                found = true;
                return "Found " + item + " at position " + (mid + 1);
            }
            else {
                right = mid - 1;
            }
        }
        return item + " not found!";
    }
    binaryRec(array, item, left, right) {
        if (left <= right) {
            let mid = Math.floor((left + right) / 2);
            if (item > array[mid]) {
                return this.binaryRec(array, item, mid + 1, right);
            }
            if (item == array[mid]) {
                return "Found " + item + " at position " + (mid + 1);
            }
            if (item < array[mid]) {
                return this.binaryRec(array, item, left, mid - 1);
            }
        }
        return item + " not found";
    }
}

class DataStructure {
    constructor() {
        if (this.constructor === DataStructure) {
            throw new Error("Abstract Class");
        }
    }
}

class Node {
    constructor(item) {
        this.item = item;
        this.left = null;
        this.right = null;
        this.front = null;
        this.rear = null;
        this.parent = null;
        this.root = null;
        this.top = null;
    }
}

class LinkedList {
    constructor() {
        this.length = 0;
        this.root = null;
    }
    isEmpty() {
        return this.root == null;
    }
    insertLast(item) {
        let newNode = new Node(item);
        if (this.isEmpty()) {
            this.root = newNode;
        }
        else {
            let current = this.root;
            while (current.right != null) {
                current = current.right;
            }
            current.right = newNode;
        }
        this.length++;
    }
    insertFirst(item) {
        let newNode = new Node(item);
        newNode.right = this.root;
        this.root = newNode;
        this.length++;
    }
    insertAt(index, item) {
        let newNode = new Node(item);
        if (this.isEmpty() || index < 2) {
            this.insertFirst(item);
            console.log("LL is empty, item is inserted at position 1");
            return;
        }
        else {
            let current = this.root;
            let i = 1;
            while (current.right != null && i < index - 1) {
                current = current.right;
                i++;
            }
            newNode.right = current.right;
            current.right = newNode;
            console.log("Item is inserted at position " + i);
        }
        this.length++;
    }
    search(item) {
        if (this.isEmpty()) {
            console.log("LL is empty");
        }
        else {
            let current = this.root;
            let i = 1;
            while (current.right != null && current.item != item) {
                current = current.right;
                i++;
            }
            if (current.item == item) {
                console.log("item found at position " + i);
                return i;
            }
            else {
                console.log("item not found");
                return -1;
            }
        }
    }
    remove(item) {
        if (this.isEmpty()) {
            console.log("LL is empty");
        }
        else {
            let current = this.root;
            let preCurrent = null;
            while (current.item != item) {
                if (current.right == null) {
                    console.log("item not found");
                    return -1;
                }
                else {
                    preCurrent = current;
                    current = current.right;
                }
            }
            if (preCurrent == null) {
                this.root = this.root.right;
                return "deleted";
            }
            else {
                preCurrent.right = current.right;
                return "deleted";
            }
            this.length--;
        }
    }
    print() {
        let current = this.root;
        while (current) {
            console.log(current.item + "\n");
            current = current.right;
        }
    }
}

class Queue {
    constructor() {
        this.length = 0;
        this.front = null;
        this.rear = null;
    }
    isEmpty() {
        return this.front == null && this.rear == null;
    }
    enqueue(item) {
        let newNode = new Node(item);
        if (this.isEmpty()) {
            this.front = newNode;
            this.rear = newNode;
        }
        else {
            this.rear.right = newNode;
            this.rear = newNode;
        }
        this.length++;
    }
    dequeue() {
        if (!this.isEmpty()) {
            if (this.front == this.rear) {
                this.front == null;
                this.rear = null;
            }
            else {
                this.front = this.front.right;
            }
            this.length--;
        }
        else {
            console.log("Queue is empty!");
        }
    }
    print() {
        let current = this.front;
        while (current) {
            console.log(current.item + "\n");
            current = current.right;
        }
    }
}

class Stack {
    constructor() {
        this.length = 0;
        this.top = null;
    }
    isEmpty() {
        return this.top == null;
    }
    push(item) {
        let newNode = new Node(item);
        newNode.right = this.top;
        this.top = newNode;
        this.length++;
    }
    pop() {
        if (!this.isEmpty()) {
            this.top = this.top.right;
            this.length--;
        }
        else {
            console.log("Stack is empty!");
        }
    }
    print() {
        let current = this.top;
        while (current) {
            console.log(current.item + "\n");
            current = current.right;
        }
    }
}

class DoubleLinkedList {
    constructor() {
        this.root = null;
        this.length = 0;
    }
    isEmpty() {
        return this.root == null;
    }
    insertFirst(item) {
        let newNode = new Node(item);
        newNode.right = this.root;
        this.root = newNode;
        this.length++;
    }
    insertLast(item) {
        let newNode = new Node(item);
        let lastNode = this.root;
        while (lastNode && lastNode.right != null) {
            lastNode = lastNode.right;
        }
        if (lastNode) {
            lastNode.right = newNode;
            newNode.left = lastNode;
        }
        else {
            this.insertFirst(item);
            console.log("DLL is empty, item is inserted at position 1");
            return;
        }
        this.length++;
    }
    insertAt(index, item) {
        let newNode = new Node(item);
        if (this.isEmpty() || index < 2) {
            this.insertFirst(item);
            console.log("Dll is empty, item is inserted at position 1");
            return;
        }
        else {
            let current = this.root;
            let i = 1;
            while (current.right != null && i < index - 1) {
                current = current.right;
                i++;
            }
            newNode.right = current.right;
            newNode.left = current;
            current.right = newNode;
            this.length++;
        }
    }
    remove(item) {
        if (this.isEmpty()) {
            console.log("DLL is empty!");
        }
        else {
            let current = this.root;
            let preCurrent = null;
            while (current.item != item) {
                if (current.right == null) {
                    console.log("item not found");
                    return -1;
                }
                else {
                    preCurrent = current;
                    current = current.right;
                }
            }
            if (preCurrent == null) {
                this.root = current.right;
                this.root.left = null;
            }
            else {
                preCurrent.right = current.right;
                current.right ? (current.right.left = preCurrent) : null;
            }
            this.length--;
        }
    }
    print() {
        let current = this.root;
        while (current) {
            console.log(current.item + "\n");
            current = current.right;
        }
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null;
        this.length = 0;
    }
    isEmpty() {
        return this.root == null;
    }
    insert(item) {
        let newNode = new Node(item);
        if (this.isEmpty()) {
            this.root = newNode;
        }
        else {
            let current = this.root;
            while (current.right != null && current.left != null) {
                if (item >= current.item) {
                    current = current.right;
                }
                else {
                    current = current.left;
                }
            }
            if (item >= current.item) {
                current.right = newNode;
            }
            else {
                current.left = newNode;
            }
            newNode.parent = current;
        }
    }
    search(item) {
        if (this.isEmpty()) {
            console.log("Tree is empty");
            return -1;
        }
        else {
            let current = this.root;
            let level = 1;
            while (current) {
                if (item > current.item) {
                    current = current.right;
                }
                else if (item == current.item) {
                    console.log(item + " found at level " + level);
                    return level;
                }
                else {
                    current = current.left;
                }
                level++;
            }
            if (current == null) {
                console.log("NOT found!");
                return -1;
            }
        }
    }
    preOrder(node) {
        if (node) {
            console.log(node.item);
            this.preOrder(node.left);
            this.preOrder(node.right);
        }
    }
    inOrder(node) {
        if (node) {
            this.inOrder(node.left);
            console.log(node.item);
            this.inOrder(node.right);
        }
    }
    postOrder(node) {
        if (node) {
            this.postOrder(node.left);
            this.postOrder(node.right);
            console.log(node.item);
        }
    }
}

let algo = new Algorithm();
let ll = new LinkedList();
let dll = new DoubleLinkedList();
let queue = new Queue();
let stack = new Stack();
let tree = new BinarySearchTree();

File: Algorithms & Data Structures/Data structures implementation C++/All Combined.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <cstdlib>
using namespace std;

struct node {
  int data;
  node *next;
  node *parent;
  node *left;
  node *right;
};

// Unsorted Linked List //

class LinkedList {
  public:
    node *start;
    LinkedList() {
      start = NULL;
    }
    void addFirst(int item) {
      node *temp = new node;
      temp->data = item;
      temp->next = start;
      start = temp;
    }
    void addAfter(int item, int after) {
      node *temp = new node;
      temp->data = item;
      node *location = start;
      while(location->data != after) {
        location = location->next;
      }
      temp->next = location->next;
      location->next = temp;
    }
    void addLast(int item) {
      node *temp = new node;
      temp->data = item;
      temp->next = NULL;
      node *location = start;
      while(location != NULL && location->next != NULL) {
        location = location->next;
      }
      if(location == NULL) {
        temp->next = location;
        start = temp;
      }
      else {
        location->next = temp;
      }
    }
    void deleteItem(int item) {
      node *location = start;
      node *preLocation = NULL;
      if(location->data == item) {
        start = start->next;
      }
      else {
        while(location->data != item) {
          preLocation = location;
          location = location->next;
        }
        preLocation->next = location->next;
        delete location;
      }
    }
    bool search(int item) {
      node *location = start;
      while(location != NULL) {
        if(location->data == item) {
          return true;
        }
        location = location->next;
      }
      return false;
    }
};
// Unsorted linked list END //

// Sorted Linked List //

class SortedLinkedList {
  public:
    node *start;
    SortedLinkedList() {
      start = NULL;
    }
    void insert(int item) {
      node *temp = new node;
      temp->data = item;
      node *location = start;
      node *preLocation = NULL;
      while(location != NULL && item > location->data) {
        preLocation = location;
        location = location->next;
      }
      if(preLocation == NULL) {
        temp->next = start;
        start = temp;
      }
      else {
        temp->next = preLocation->next;
        preLocation->next = temp;
      }
    }
    void deleteItem(int item) {
      node *location = start;
      node *preLocation = NULL;
      if(location->data == item) {
        start = start->next;
      }
      else {
        while(location != NULL && item > location->data) {
          preLocation = location;
          location = location->next;
        }
        preLocation->next = location->next;
        delete location;
      }
    }
    bool search(int item) {
      node *location = start;
      while(location != NULL && item >= location->data) {
        if(location->data == item) {
          return true;
        }
        location = location->next;
      }
      return false;
    }
};

// Sorted Linked List END //

// Queue //

class queue {
  public:
    node *front, *rear;
    queue() {
      front = NULL;
      rear = NULL;
    }
    void enqueue(int item) {
      node *temp = new node;
      temp->data = item;
      temp->next = NULL;
      if(front == NULL) {
        front = temp;
        rear = temp;
      }
      else {
        rear->next = temp;
        rear = temp;
      }
    }
    int dequeue() {
      node *old = front;
      front = front->next;
      if(front == NULL) {
        rear = NULL; // I'm not sure of importance of this line :D //
      }
      int oldData = old->data;
      delete old;
      return oldData;
    }
};

// Queue END //


// Stack //

class stack {
  public:
    node *top;
    stack() {
      top = NULL;
    }
    void push(int item) {
      node *temp = new node;
      temp->data = item;
      temp->next = top;
      top = temp;
    }
    int pop() {
      node *old = top;
      top = top->next;
      int oldData = old->data;
      delete old;
      return oldData;
    }
};

//Stack END //

//Tree //
class tree {
    public:
        node *root;
        tree() {
            root = NULL;
        }
        bool isEmpty() {
            return (root == NULL);
        }
        void insert(int item) {
            node *temp = new node;
            temp->data = item;
            temp->left = NULL;
            temp->right = NULL;
            if(isEmpty()) {
                temp->parent = NULL;
                root = temp;
            }
            else {
                node *location=root, *preLocation=NULL;
                while(location!=NULL) {
                    if(item > location->data) {
                        preLocation = location;
                        location = location->right;
                    }
                    else {
                        preLocation = location;
                        location = location->left;
                    }
                }
                if(item > preLocation->data) {
                    preLocation->right = temp;
                }
                else {
                    preLocation->left = temp;
                }
                temp->parent = preLocation;
            }
        }
        bool search(int item) {
            node *location = root;
            while(location != NULL) {
                if(item > location->data) {
                    location = location->right;
                }
                else if(item < location->data) {
                    location = location->left;
                }
                else {
                    return true;
                }
            }
            return false;
        }
        void deleteItem(int item) {
            node *location=root, *preLocation=NULL;
            while(location != NULL) {
                if(item > location->data) {
                    preLocation = location;
                    location = location->right;
                }
                else if(item < location->data) {
                    preLocation = location;
                    location = location->left;
                }
                else if(item == location->data) {
                    break;
                }
            }
            if(location == NULL) {
                cout << "element not found" << endl;
            }
            else if(location->left != NULL && location->right != NULL) {
                node *successor = location->right;
                while(successor->left != NULL) {
                    successor = successor->left;
                }
                if(successor->right != NULL) {
                    if(successor->right->data > location->data) {
                      location->right = successor->right;
                      successor->right->parent = location;
                    }
                    else {
                        successor->parent->left = successor->right;
                        successor->right->parent = successor->parent;
                    }
                }
                if(successor->data > successor->parent->data) {
                  successor->parent->right = NULL;
                }
                else {
                  successor->parent->left = NULL;
                }
                int temp = successor->data;
                successor->data = location->data;
                location->data = temp;
                location = successor;
            }
            else if(location->left != NULL || location->right != NULL) {
                if(location->right != NULL && location->right->data > preLocation->data) {
                    preLocation->right = location->right;
                    location->right->parent = preLocation;
                }
                else if(location->right != NULL && location->right->data < preLocation->data) {
                    preLocation->left = location->right;
                    location->right->parent = preLocation;
                }
                else if(location->left != NULL && location->left->data > preLocation->data) {
                    preLocation->right = location->left;
                    location->left->parent = preLocation;
                }
                else {
                    preLocation->left = location->left;
                    location->left->parent = preLocation;
                }
            }
            else {
                if(location->data > preLocation->data) {
                    preLocation->right = NULL;
                }
                else {
                    preLocation->left = NULL;
                }
            }
            delete location;
        }
        void breadthFirst() {
            node *left=root->left, *right=root->right;
            cout << root->data << "\t";
            while(left != NULL || right != NULL) {
                if(left != NULL) {
                    cout << left->data << "\t";
                    if(right == NULL) {
                        right = left->right;
                    }
                    left = left->left;
                }
                if(right != NULL) {
                    cout << right->data << "\t";
                    if(left == NULL) {
                        left = right->left;
                    }
                    right = right->right;
                }
            }
        }
        void preOrder(node *direction) {
            if(direction) {
                cout << direction->data << "\t";
                preOrder(direction->left);
                preOrder(direction->right);
            }
        }
        void inOrder(node *direction) {
            if(direction) {
                inOrder(direction->left);
                cout << direction->data << " ";
                inOrder(direction->right);
            }
        }
        void postOrder(node *direction) {
            if(direction) {
                postOrder(direction->left);
                postOrder(direction->right);
                cout << direction->data << "\t";
            }
        }
};

// Tree END //

// Function to print items of (linked-list, queue, stack)

void print(node *location) {
  while(location != NULL) {
    cout << location->data << endl;
    location = location->next;
  }
}
// print function END //


int main() {
 // Linked-list: addFirst, addAfter, addLast, deleteItem, search
 LinkedList l;
 // Unsorted LL: insert, search, deleteItem
 SortedLinkedList ll;
 // Queue: enqueue, dequeue
 queue q;
 // Stack: push, pop
 stack s;
 // Tree: insert, search, pre/in/post - order, deleteItem, breadthFirst
 tree t;
}